arangod/Cluster/AgencyCache.cpp

```diff
@@ -469,3 +567,129 @@ void AgencyCache::invokeAllCallbacks() const {
   }
   invokeCallbacks(toCall);
 }
+
+
+AgencyCache::change_set_t AgencyCache::changedSince(
+  std::string const& what, consensus::index_t const& last) const {
+
+  static std::vector<std::string> const planGoodies ({
+      AgencyCommHelper::path(PLAN_ANALYZERS) + "/",
+      AgencyCommHelper::path(PLAN_COLLECTIONS) + "/",
+      AgencyCommHelper::path(PLAN_DATABASES) + "/",
+      AgencyCommHelper::path(PLAN_VIEWS) + "/"});
+  static std::vector<std::string> const currentGoodies ({
+      AgencyCommHelper::path(CURRENT_COLLECTIONS) + "/",
+      AgencyCommHelper::path(CURRENT_DATABASES) + "/"});
+
+  bool get_rest = false;
+  std::unordered_map<std::string, query_t> db_res;
+  query_t rest_res = nullptr;
+
+  decltype(_planChanges) const& changes = (what == PLAN) ? _planChanges : _currentChanges;
+  std::vector<std::string> const& goodies = (what == PLAN) ? planGoodies : currentGoodies;
+
+  std::unordered_set<std::string> databases;
+  std::shared_lock g(_storeLock);
+
+  auto tmp = _readDB.nodePtr()->hasAsUInt(AgencyCommHelper::path(what) + "/" + VERSION);
+  uint64_t version = tmp.second ? tmp.first : 0;
+
+  if (last < _lastSnapshot) {
+    get_rest = true;
+    auto keys = _readDB.nodePtr(AgencyCommHelper::path(what) + "/" + DATABASES)->keys();
+    for (auto& i : keys) {
+      databases.emplace(std::move(i));
```

Comment:
why not to reserve space for all keys before entering the loop?

Reply:
sure. good point